IN predicate for strings

oerling opened the pull request:
Adds BytesValues for efficient set membership filter for a string column.
Uses a linear hash table and Bloom filter for set membership checking and works without allocating memory.
Adds a low-cost hash function over a byte[] that bypasses using Slice for passing a string v iew and uses a cheaper hash function than XxHash64 from Slice.

tdcmeehan commented on presto-orc/src/main/java/com/facebook/presto/orc/TupleDomainFilter.java:
```diff
@@ -839,6 +850,137 @@ public String toString()
+        private final byte[][] values;
+        private final byte[][] hashTable;
+        private final int hashTableSizeMask;
+        long[] bloom;
```
I think these should have private visibility.

mbasmanova replied:
bloom and bloomSize need to be private final

mbasmanova commented on presto-orc/src/main/java/com/facebook/presto/orc/TupleDomainFilterUtils.java:
```diff
@@ -102,6 +103,19 @@ public static TupleDomainFilter toFilter(Domain domain)
+                return BytesValues.of(
+                        bytesRanges.stream()
+                                .map(BytesRange::getLower)
+                        .toArray(byte[][]::new),
```
wrong indentation

mbasmanova commented on presto-orc/src/main/java/com/facebook/presto/orc/TupleDomainFilter.java:
```diff
@@ -839,6 +850,137 @@ public String toString()
+        long[] bloom;
+        int bloomSize;
+
+        public BytesValues(byte[][] values, boolean nullAllowed)
```
BytesValues constructor needs to be private

mbasmanova commented on presto-orc/src/main/java/com/facebook/presto/orc/TupleDomainFilter.java:
```diff
@@ -839,6 +850,137 @@ public String toString()
+
+            this.values = values;
+            // Linear hash table size is the next power of two greater
+            // than or equal to number of values* 5.
```
space is missing between values and *

mbasmanova commented on presto-orc/src/main/java/com/facebook/presto/orc/TupleDomainFilter.java:
```diff
@@ -839,6 +850,137 @@ public String toString()
+            bloom = new long[bloomSize];
+            for (byte[] value : values) {
+                long hashCode = hash(value, 0, value.length);
+                addBloom(hashCode);
```
I'd inline this method. It is used only once and it is just one line of implementation.

mbasmanova commented on presto-orc/src/main/java/com/facebook/presto/orc/TupleDomainFilter.java:
```diff
@@ -839,6 +850,137 @@ public String toString()
+            if (values.length != that.values.length || nullAllowed != that.nullAllowed) {
+                return false;
+            }
+            // Arrays.equals is not recursive: Compare a byte[] at a time.
```
I think you can use Objects.deepEquals

mbasmanova commented on presto-orc/src/main/java/com/facebook/presto/orc/ByteArrayUtils.java:
```diff
@@ -83,4 +83,20 @@ private static long longBytesToLong(long bytes)
         return Long.reverseBytes(bytes) ^ Long.MIN_VALUE;
     }
+
+    public static long hash(byte[] bytes, int offset, int length)
```
Does this hash algorithm has a name? If so, let's move this method into a separate class named after that algorithm. Otherwise, where does this algorithm come from?

mbasmanova commented on presto-orc/src/main/java/com/facebook/presto/orc/ByteArrayUtils.java:
```diff
@@ -83,4 +83,20 @@ private static long longBytesToLong(long bytes)
+    public static long hash(byte[] bytes, int offset, int length)
+    {
+        // Constant from MurMur hash.
+        final long m = 0xc6a4a7935bd1e995L;
```
We don't use final for local variables. This needs to be private static final member variable.

oerling commented:
If Java had template instantiation the Bloom filter and hash table would be a template. In the present case we have this kind of structure only in this place. So having this as a utility is not Indicated. The other use case for this sort of thing is prefilter of a hash join. This is a different loop. See the initial Aria hash join for the use case. The uses are analogous but one would still need a different member function into the hash table class. So there would be no savings in lines of code and readability would be worse. And if you needed to pass callbacks for comparison/hash number calculation you would be liable to unpredictable behavior when use cases moved away from monomorphic. So having this inline is more tractable.

 

 

From: Timothy Meehan <notifications@github.com> 
Sent: Friday, July 12, 2019 9:54 AM
To: prestodb/presto <presto@noreply.github.com>
Cc: oerling <erling@xs4all.nl>; Author <author@noreply.github.com>
Subject: Re: [prestodb/presto] IN predicate for strings (#13078)

 

@tdcmeehan approved this pull request.

This is sort of generic question, but I wonder if the hash table and bloom filter can be extracted out into a separate utility.

  _____  

In presto-orc/src/main/java/com/facebook/presto/orc/ByteArrayUtils.java <https://github.com/prestodb/presto/pull/13078#discussion_r303044471> :

> @@ -83,4 +83,20 @@ private static long longBytesToLong(long bytes)
     {
         return Long.reverseBytes(bytes) ^ Long.MIN_VALUE;
     }
+
+    public static long hash(byte[] bytes, int offset, int length)
+    {
+        // Constant from MurMur hash.
+        final long m = 0xc6a4a7935bd1e995L;

I would make it a private static member.

  _____  

In presto-orc/src/main/java/com/facebook/presto/orc/TupleDomainFilter.java <https://github.com/prestodb/presto/pull/13078#discussion_r303059826> :

> @@ -839,6 +850,137 @@ public String toString()
         }
     }
 
+    public static class BytesValues
+            extends AbstractTupleDomainFilter
+    {
+        private final byte[][] values;
+        private final byte[][] hashTable;
+        private final int hashTableSizeMask;
+        long[] bloom;

I think these should have private visibility.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub <https://github.com/prestodb/presto/pull/13078?email_source=notifications&email_token=AKPPPTY6KILGRURUAEP7DU3P7CZIVA5CNFSM4H7SOPF2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB6J3ZWA#pullrequestreview-261340376> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AKPPPT3CXVAHHVJ57GAUW6LP7CZIVANCNFSM4H7SOPFQ> .  <https://github.com/notifications/beacon/AKPPPT5MYDX7CPAVIO37UJ3P7CZIVA5CNFSM4H7SOPF2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB6J3ZWA.gif>

oerling commented:
The hash is a hybrid of Knuth multiplicative and Murmur. The reasonable minimum for a hash is multiplication, shift and xor.

Murmur is about 3x the arithmetic, which is overkill for the purpose.  I don’t know of a common name for this. Added a comment.

 

Addressed the other review comments.

Added another test.

 

 

From: Maria Basmanova <notifications@github.com> 
Sent: Tuesday, July 16, 2019 6:24 AM
To: prestodb/presto <presto@noreply.github.com>
Cc: oerling <erling@xs4all.nl>; Mention <mention@noreply.github.com>
Subject: Re: [prestodb/presto] IN predicate for strings (#13078)

 

@mbasmanova commented on this pull request.

@oerling <https://github.com/oerling>  Some further comments.

  _____  

In presto-orc/src/main/java/com/facebook/presto/orc/ByteArrayUtils.java <https://github.com/prestodb/presto/pull/13078#discussion_r303904886> :

> @@ -83,4 +83,20 @@ private static long longBytesToLong(long bytes)
     {
         return Long.reverseBytes(bytes) ^ Long.MIN_VALUE;
     }
+
+    public static long hash(byte[] bytes, int offset, int length)

Does this hash algorithm has a name? If so, let's move this method into a separate class named after that algorithm. Otherwise, where does this algorithm come from?

  _____  

In presto-orc/src/main/java/com/facebook/presto/orc/ByteArrayUtils.java <https://github.com/prestodb/presto/pull/13078#discussion_r303905196> :

> @@ -83,4 +83,20 @@ private static long longBytesToLong(long bytes)
     {
         return Long.reverseBytes(bytes) ^ Long.MIN_VALUE;
     }
+
+    public static long hash(byte[] bytes, int offset, int length)
+    {
+        // Constant from MurMur hash.
+        final long m = 0xc6a4a7935bd1e995L;

We don't use final for local variables. This needs to be private static final member variable.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/prestodb/presto/pull/13078?email_source=notifications&email_token=AKPPPT2N4DM46DHCC47QMRTP7XDVRA5CNFSM4H7SOPF2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB6SBEOI#pullrequestreview-262410809> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AKPPPTYQCTOJEHQZ2YKYY6LP7XDVRANCNFSM4H7SOPFQ> .

mbasmanova commented on presto-orc/src/main/java/com/facebook/presto/orc/ByteArrayUtils.java:
```diff
@@ -83,4 +86,21 @@ private static long longBytesToLong(long bytes)
         return Long.reverseBytes(bytes) ^ Long.MIN_VALUE;
     }
+
+    public static long hash(byte[] bytes, int offset, int length)
```
`hash` is a very generic name; do you expect to use this kind of hash somewhere else? if not, perhaps, move it into BytesValues class.

mbasmanova commented on presto-orc/src/main/java/com/facebook/presto/orc/TupleDomainFilter.java:
```diff
@@ -839,6 +851,123 @@ public String toString()
         }
     }
 
+    public static class BytesValues
```
public static are redundant here; please, remove

mbasmanova commented on presto-orc/src/main/java/com/facebook/presto/orc/TupleDomainFilter.java:
```diff
@@ -839,6 +851,123 @@ public String toString()
+            if (!testBloom(hashCode)) {
+                return false;
+            }
+            int position = (int) (hashCode & (hashTableSizeMask));
```
unnecessary parentheses around hashTableSizeMask: `int position = (int) (hashCode & hashTableSizeMask);`

mbasmanova commented on presto-orc/src/main/java/com/facebook/presto/orc/TupleDomainFilter.java:
```diff
@@ -839,6 +851,123 @@ public String toString()
+            }
+            int position = (int) (hashCode & (hashTableSizeMask));
+            for (int i = position; i <= position + hashTableSizeMask; i++) {
+                int index = i & (hashTableSizeMask);
```
unnecessary parentheses around hashTableSizeMask: int index = i & hashTableSizeMask;

mbasmanova commented on presto-orc/src/main/java/com/facebook/presto/orc/TupleDomainFilter.java:
```diff
@@ -839,6 +851,123 @@ public String toString()
+            checkArgument(values.length > 1, "values must contain at least 2 entries");
+
+            this.values = values;
+            // Linear hash table size is the next power of two greater
```
- these comment lines are too short; the whole comment fits into a single line; would you make it so?
- why "times 5"?

mbasmanova commented on presto-orc/src/main/java/com/facebook/presto/orc/TupleDomainFilter.java:
```diff
@@ -839,6 +851,123 @@ public String toString()
+            int hashTableSize = Integer.highestOneBit(values.length * 5);
+            hashTableSizeMask = hashTableSize - 1;
+            hashTable = new byte[hashTableSize][];
+            // 8 bits of Bloom filter per slot in hash table. The
```
these comment lines are two short; the whole comment fits into a single line; would you make it so?

oerling commented:
Thanks, I addressed the comments.

 

The name ByteArrayUtils.hash is appropriate since the namespace gives the data type.  The intended other uses for this are in places like hash join and pushed down hash join Bloom filter. ByteArrayUtils belongs more properly in spi, as it is in  Aria research branches since the users are scan and diverse operators. We’ll get to relocating or duplicating this yet but there is no hurry for this right now.

 

 

 

Basmanova <notifications@github.com> 
Sent: Wednesday, July 17, 2019 7:37 AM
To: prestodb/presto <presto@noreply.github.com>
Cc: oerling <erling@xs4all.nl>; Mention <mention@noreply.github.com>
Subject: Re: [prestodb/presto] IN predicate for strings (#13078)

 

@mbasmanova requested changes on this pull request.

@oerling <https://github.com/oerling>  Orri, thank you for the updated and for adding a new test. I have some followup comments and I see a test failure.

[ERROR] testBytesValues(com.facebook.presto.orc.TestTupleDomainFilter)  Time elapsed: 0.024 s  <<< FAILURE!
java.lang.ArrayIndexOutOfBoundsException: 111
  at com.facebook.presto.orc.TestTupleDomainFilter.testBytesValues(TestTupleDomainFilter.java:278)
  at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
  at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
  at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
  _____  


In presto-orc/src/main/java/com/facebook/presto/orc/TupleDomainFilter.java <https://github.com/prestodb/presto/pull/13078#discussion_r304159641> :

> +            return mask == (bloom[index] & mask);
+        }
+
+        @Override
+        public boolean equals(Object o)
+        {
+            if (this == o) {
+                return true;
+            }
+
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+
+            BytesValues that = (BytesValues) o;
+            return values.length == that.values.length && nullAllowed == that.nullAllowed && deepEquals(values, that.values);

*	values.length == that.values.length is redundant as deepEquals includes this check
*	qualify deepEquals for consistency with Arrays.equals elsewhere: Arrays.deepEquals
*	would you put each conjunct on a separate line?

            return nullAllowed == that.nullAllowed &&
                    Arrays.deepEquals(values, that.values);
  _____  


In presto-orc/src/main/java/com/facebook/presto/orc/ByteArrayUtils.java <https://github.com/prestodb/presto/pull/13078#discussion_r304169218> :

> @@ -23,6 +23,9 @@
 
 public class ByteArrayUtils
 {
+        // Constant from MurMur hash.

wrong indentation

  _____  

In presto-orc/src/main/java/com/facebook/presto/orc/ByteArrayUtils.java <https://github.com/prestodb/presto/pull/13078#discussion_r304421378> :

> @@ -83,4 +86,21 @@ private static long longBytesToLong(long bytes)
     {
         return Long.reverseBytes(bytes) ^ Long.MIN_VALUE;
     }
+
+    public static long hash(byte[] bytes, int offset, int length)

hash is a very generic name; do you expect to use this kind of hash somewhere else? if not, perhaps, move it into BytesValues class.

  _____  

In presto-orc/src/main/java/com/facebook/presto/orc/TupleDomainFilter.java <https://github.com/prestodb/presto/pull/13078#discussion_r304422648> :

> @@ -839,6 +851,123 @@ public String toString()
         }
     }
 
+    public static class BytesValues

public static are redundant here; please, remove

  _____  

In presto-orc/src/main/java/com/facebook/presto/orc/TupleDomainFilter.java <https://github.com/prestodb/presto/pull/13078#discussion_r304436040> :

> +            }
+        }
+
+        public static BytesValues of(byte[][] values, boolean nullAllowed)
+        {
+            return new BytesValues(values, nullAllowed);
+        }
+
+        @Override
+        public boolean testBytes(byte[] value, int offset, int length)
+        {
+            long hashCode = hash(value, offset, length);
+            if (!testBloom(hashCode)) {
+                return false;
+            }
+            int position = (int) (hashCode & (hashTableSizeMask));

unnecessary parentheses around hashTableSizeMask: int position = (int) (hashCode & hashTableSizeMask);

  _____  

In presto-orc/src/main/java/com/facebook/presto/orc/TupleDomainFilter.java <https://github.com/prestodb/presto/pull/13078#discussion_r304436381> :

> +
+        public static BytesValues of(byte[][] values, boolean nullAllowed)
+        {
+            return new BytesValues(values, nullAllowed);
+        }
+
+        @Override
+        public boolean testBytes(byte[] value, int offset, int length)
+        {
+            long hashCode = hash(value, offset, length);
+            if (!testBloom(hashCode)) {
+                return false;
+            }
+            int position = (int) (hashCode & (hashTableSizeMask));
+            for (int i = position; i <= position + hashTableSizeMask; i++) {
+                int index = i & (hashTableSizeMask);

unnecessary parentheses around hashTableSizeMask: int index = i & hashTableSizeMask;

  _____  

In presto-orc/src/main/java/com/facebook/presto/orc/TupleDomainFilter.java <https://github.com/prestodb/presto/pull/13078#discussion_r304437681> :

> +    {
+        private final byte[][] values;
+        private final byte[][] hashTable;
+        private final int hashTableSizeMask;
+        private final long[] bloom;
+        private final int bloomSize;
+
+        private BytesValues(byte[][] values, boolean nullAllowed)
+        {
+            super(nullAllowed);
+
+            requireNonNull(values, "values is null");
+            checkArgument(values.length > 1, "values must contain at least 2 entries");
+
+            this.values = values;
+            // Linear hash table size is the next power of two greater

*	these comment lines are too short; the whole comment fits into a single line; would you make it so?
*	why "times 5"?

  _____  

In presto-orc/src/main/java/com/facebook/presto/orc/TupleDomainFilter.java <https://github.com/prestodb/presto/pull/13078#discussion_r304437738> :

> +        private final int bloomSize;
+
+        private BytesValues(byte[][] values, boolean nullAllowed)
+        {
+            super(nullAllowed);
+
+            requireNonNull(values, "values is null");
+            checkArgument(values.length > 1, "values must contain at least 2 entries");
+
+            this.values = values;
+            // Linear hash table size is the next power of two greater
+            // than or equal to number of values * 5.
+            int hashTableSize = Integer.highestOneBit(values.length * 5);
+            hashTableSizeMask = hashTableSize - 1;
+            hashTable = new byte[hashTableSize][];
+            // 8 bits of Bloom filter per slot in hash table. The

these comment lines are two short; the whole comment fits into a single line; would you make it so?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/prestodb/presto/pull/13078?email_source=notifications&email_token=AKPPPT6HIJFR47VFZHMZBSTP74VCJA5CNFSM4H7SOPF2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB6UPD2Q#pullrequestreview-262730218> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AKPPPT2QLUESUCLQNXDGGQ3P74VCJANCNFSM4H7SOPFQ> .  <https://github.com/notifications/beacon/AKPPPT2LX5EEWFUKJO55FSTP74VCJA5CNFSM4H7SOPF2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB6UPD2Q.gif>
